File: python/bican_mccarroll_eqtl/src/bican_mccarroll_eqtl/fonts.py
"""Manuscript font handling for matplotlib-based plots.

Mirrors the R-side ``fonts.R`` in ``bican.mccarroll.eqtl``: a single place
that pins the font used for manuscript figures, and a loud check for when
that font is not actually resolvable on the current system.
"""

import warnings
from contextlib import contextmanager

import matplotlib as mpl
from matplotlib import font_manager

#: Font used for all manuscript figures.
MANUSCRIPT_FONT = "Arial"


@contextmanager
def manuscript_font():
    """Context manager that renders matplotlib text in the manuscript font.

    Sets ``font.family``/``font.sans-serif`` to :data:`MANUSCRIPT_FONT` and
    ``svg.fonttype`` to ``"none"`` so that SVG output keeps real ``<text>``
    elements (editable, with the font name embedded) rather than converting
    glyphs to vector outlines. Settings are restored on exit via
    ``matplotlib.rc_context``, even if an exception is raised.
    """
    with mpl.rc_context(
        {
            "font.family": "sans-serif",
            "font.sans-serif": [MANUSCRIPT_FONT],
            "svg.fonttype": "none",
        }
    ):
        yield


def check_manuscript_font():
    """Warn if the manuscript font is not resolvable on this system.

    matplotlib silently substitutes a fallback font if
    :data:`MANUSCRIPT_FONT` is not installed, so this check exists to
    surface that condition loudly instead of letting figures regenerate in
    the wrong font unnoticed.
    """
    try:
        found_path = font_manager.findfont(
            MANUSCRIPT_FONT, fallback_to_default=False
        )
    except Exception:
        found_path = None

    if found_path is None or MANUSCRIPT_FONT.lower() not in found_path.lower():
        fallback = found_path or font_manager.findfont(MANUSCRIPT_FONT)
        warnings.warn(
            f"Manuscript font '{MANUSCRIPT_FONT}' not found on this system; "
            f"figures will fall back to '{fallback}'.",
            stacklevel=2,
        )

File: R/bican.mccarroll.eqtl/R/fonts.R
#' Font used for all manuscript figures.
MANUSCRIPT_FONT <- "Arial"

#' Open an SVG device that renders text in the manuscript font
#'
#' Wraps \code{svglite::svglite()}, binding \code{MANUSCRIPT_FONT} to the
#' \code{"sans"} font alias so that ggplot2's default (unset) text family
#' resolves to it. Suitable both as a direct device opener
#' (\code{svglite_manuscript(file, width, height)}) and as a \code{ggsave}
#' \code{device} argument.
#'
#' @param filename Output SVG file path.
#' @param ... Additional arguments passed to \code{svglite::svglite()}.
#' @return Invisibly, the result of \code{svglite::svglite()}.
svglite_manuscript <- function(filename, ...) {
  svglite::svglite(
    filename,
    ...,
    system_fonts = list(sans = MANUSCRIPT_FONT)
  )
}

#' Warn if the manuscript font is not resolvable on this system
#'
#' Called at package load. \code{svglite_manuscript()} silently falls back to
#' a substitute font if \code{MANUSCRIPT_FONT} is not installed, so this check
#' exists to surface that condition loudly instead of letting figures
#' regenerate in the wrong font unnoticed.
.check_manuscript_font <- function() {
  match <- systemfonts::match_fonts(MANUSCRIPT_FONT)
  if (!grepl(MANUSCRIPT_FONT, basename(match$path), ignore.case = TRUE)) {
    warning(
      sprintf(
        "Manuscript font '%s' not found on this system; figures will fall back to '%s'.",
        MANUSCRIPT_FONT,
        basename(match$path)
      ),
      call. = FALSE
    )
  }
}
